In [343]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import math
import scipy.stats as stats
import pandas as pd

# 1. Statatistical Test

In [344]:
class Stats_Test():    
        
    def Test(self, array_x, alternative, array_y = None, val=None, var = None, normal = False, alpha = None):
        
        ## Part 1
        #Normal 일 때
        if bool(np.mean(array_y == None)) == True and val == None:
            print("The test cannot be conducted due to empty value.")
        elif bool(np.mean(array_y == None)) == True and val != None:
            if normal == True:
                #Variance가 주어지지 않았을 때(Sample)
                if var == None:
                    Ssquare=np.var(array_x)*len(array_x)/(len(array_x)-1)
                    S=np.sqrt(Ssquare)
                    T=(np.mean(array_x)-val)/S * np.sqrt(len(array_x))
                    self.value = T
                    self.test = 'T'
                    self.df = len(array_x) -1
                #Variance가 주어졌을 때
                else:
                    Z=(np.mean(array_x)-val)/var * np.sqrt(len(array_x))
                    self.value = Z
                    self.test = 'Z'
            #Normal이 아닐 때(Central Limit Theorem)
            else:
                #Variance가 주어지지 않았을 때(Sample)
                if var == None:
                    Ssquare=np.var(array_x)*len(array_x)/(len(array_x)-1)
                    S=np.sqrt(Ssquare)
                    T=(np.mean(array_x)-val)/S * np.sqrt(len(array_x))
                    self.value = T
                    self.test = 'Z'

                #Variance가 주어졌을 때    
                else:
                    
                    Z=(np.mean(array_x)-val)/var * np.sqrt(len(array_x))
                    self.value = Z
                    self.test = 'Z'
                    
                if len(array_x)<30:
                    print("Warning: The sample size is too small.")

            if alternative == 'neq':
                if self.test == 'Z':
                    return 2*(1-stats.norm.cdf(abs(self.value)))
                if self.test == 'T':
                    return 2*(1-stats.t.cdf(abs(self.value), df = self.df))
            elif alternative == '<':
                if self.test == 'Z':
                    return stats.norm.cdf(self.value)
                if self.test == 'T':
                    return stats.t.cdf(self.value, df = self.df)
            elif alternative == '>':
                if self.test == 'Z':
                    return 1-stats.norm.cdf(self.value)
                if self.test == 'T':
                    return 1-stats.t.cdf(self.value, df = self.df)

        
        
        else:

            if val == None:
                val = 0

            if normal == True:
                #Variance가 주어지지 않았을 때(Sample)
                if var == None:
                    Ssquare1 = np.var(array_x)*len(array_x)/(len(array_x)-1)
                    S1 = np.sqrt(Ssquare1)
                    Ssquare2 = np.var(array_y)*len(array_y)/(len(array_y)-1)
                    S2 = np.sqrt(Ssquare2)
                    df1 = (Ssquare1/len(array_x)+Ssquare2/len(array_y))**2
                    df2 = (Ssquare1/len(array_x))**2/(len(array_x)-1)+ \
                    (Ssquare2/len(array_y))**2/(len(array_y)-1)
                    denominator = np.sqrt(Ssquare1/len(array_x)+Ssquare2/len(array_y))
                    T = (np.mean(array_x) - np.mean(array_y) - val)/denominator
                    self.value = T
                    self.test = 'T'
                    self.df = df1/df2
                    
                #Variance가 주어졌을 때
                else:
                    denominator = np.sqrt(var[0]**2/len(array_x)+var[1]**2/len(array_y))
                    self.test = 'Z'
                    Z = (np.mean(array_x) - np.mean(array_y) - val)/denominator
                    self.value = Z
            else:
                #Normal이 아닐 때(Central Limit Theorem)
                if len(array_x)<30 or len(array_y)<30:
                    print('Warning: The sample sizes are too small.')
                #Variance가 주어지지 않았을 때(Sample)
                if var == None:
                    Ssquare1 = np.var(array_x)*len(array_x)/(len(array_x)-1)
                    S1 = np.sqrt(Ssquare1)
                    Ssquare2 = np.var(array_y)*len(array_y)/(len(array_y)-1)
                    S2 = np.sqrt(Ssquare2)
                    denominator = np.sqrt(S1**2/len(array_x)+S2**2/len(array_y))
                    T = (np.mean(array_x) - np.mean(array_y) - val)/denominator
                    self.value = T
                    self.test = 'Z'

                #Variance가 주어졌을 때
                else:
                    Z = (np.mean(array_x)-np.mean(array_y)-val )/np.sqrt(var[0]/len(array_x)+var[1]/len(array_y))
                    self.value = Z
                    self.test = 'Z'
                
            if alternative == 'neq':
                if self.test == 'Z':
                    return 2*(1-stats.norm.cdf(abs(self.value)))
                if self.test == 'T':
                    return 2*(1-stats.t.cdf(abs(self.value), df = self.df))
            elif alternative == '<':
                if self.test == 'Z':
                    return stats.norm.cdf(self.value)
                if self.test == 'T':
                    return stats.t.cdf(self.value, df =  self.df)
            elif alternative == '>':
                if self.test == 'Z':
                    return 1-stats.norm.cdf(self.value)
                if self.test == 'T':
                    return 1-stats.t.cdf(self.value, df = self.df)

## Check it!

In [345]:
A=Stats_Test()

In [346]:
samp=np.array([7.65, 7.60, 7.65, 7.70, 7.55, 7.55, 7.40, 7.40, 7.50, 7.50])

In [347]:
A.Test(array_x=samp, normal=True, alternative="<")

The test cannot be conducted due to empty value.


In [348]:
A.Test(array_x=samp, val=7.8, normal=True, alternative="<")

1.5079814004897984e-05

In [349]:
A.Test(array_x=samp, val=7.8, normal=False, alternative="<")

7.083697760973414e-15

In [350]:
samp2=np.array([7.35, 7.68, 8.65, 5.70, 7.50, 6.75, 4.20, 5.73, 3.80, 6.40])

In [351]:
A.Test(array_x=samp, array_y=samp2 , normal=True, alternative=">", alpha=0.05)

0.01988673194901347

In [352]:
A.Test(array_x=samp, array_y=samp2 , normal=False, alternative=">")

0.008233104348804754

In [353]:
A.Test(array_x=samp, array_y=samp2 , var=[0.01,2.1], normal=False, alternative=">")

0.005297199310786649

In [354]:
A.Test(array_x=samp, array_y=samp2 , val=0.5, var=[0.01,2.1],normal=False, alternative=">")

0.07114734608030493

# 2. Linear Regression

In [355]:
# Background Functions

def householder(u,k):
    w=np.zeros(len(u))
    w[(k+1):]=u[(k+1):]
    s=np.sqrt(np.dot(u[k:],u[k:]))
    w[k]=u[k]+np.sign(u[k])*s
    length=np.sqrt(2*s*(s+np.abs(u[k])))
    if length!=0:
        w=w/length
    H=np.eye(len(u))-2*np.outer(w,w)
    return H

def forward_substitution(L,v):
    
    n=len(L)
    
    x=np.zeros(n)
    
    for i in range(n):
        tmp = v[i]-np.dot(L[i][:i],x[:i])
        x[i] = tmp/L[i,i]
        
    return x

def back_substitution(U,v):
    
    n=len(U)
    
    x = np.zeros(n)

    for i in range(n-1, -1, -1):
        tmp = v[i]-np.dot(U[i][(i+1):],x[(i+1):])
        x[i] = tmp/U[i,i]
    
    return x

def QRdecomp1(A):
    n,m=A.shape
    R=A.copy()
    Q=np.eye(n)
    for j in range(m):
        H=householder(R[:,j],j)
        Q=np.dot(Q,H)
        R=np.dot(H,R)
    return (R,Q)

def QRdecomp2(A,z):
    n,m=A.shape
    R=A.copy()
    z_=z.copy()
    #Q=np.eye(n)
    for j in range(m):
        H=householder(R[:,j],j)
        #Q=np.dot(Q,H)
        R=np.dot(H,R)
        z_=np.dot(H,z_)
    return (R[:m],z_[:m])

In [356]:
whitedata = pd.read_csv("white.csv")   # dataframe
whitedata.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [357]:
whitedata = whitedata.dropna(axis=0)

X = np.array(whitedata.drop(["quality"],1))
X = np.hstack([np.ones(len(X)).reshape(len(X),1),X])
y = np.array(whitedata["quality"])



$ X^TX = R^T R =R_1^TR_1$

where $R_1$ is upper p+1 rows of $R$

In [358]:
R = QRdecomp1(X)[0]
R1 = R[0:np.shape(X)[1],:]
XTX = np.dot(np.transpose(R1),R1)

$ (X^TX)^{-1}=(R_1^TR_1)^{-1} = (R_1)^{-1}(R_1^T)^{-1} = (R_1)^{-1}(R_1^{-1})^T$

In [359]:
#Finding R1
R = QRdecomp1(X)[0]
R1 = R[0:np.shape(X)[1],:]
XTX = np.dot(np.transpose(R1),R1)
R1I = []

# Back substitution for finding inverse of R1
R1I = []
for i in range(np.shape(R1)[1]):
    I = np.eye(np.shape(R1)[0])[:,i]
    v = back_substitution(R1,I)
    R1I.append(v)
R1I = np.transpose(R1I)
# inverse of XTX
XTXI = np.dot(R1I,np.transpose(R1I))

In [360]:
R_1,QY_1=QRdecomp2(X,y)
beta=back_substitution(R_1,QY_1)

## Solution

(a) $H_0:\beta_1 = \beta_2 = \cdots \beta_p$ vs    $H_1: \beta_i \neq 0$ for some $i$

$F = \frac{MSR}{MSE} \sim F(p,N-p-1)$

I will execute the test with $\alpha = 0.05$

In [362]:
alpha = 0.05

In [363]:
yhat = np.dot(X,np.transpose(beta))
SSE = np.dot(y - yhat,y - yhat)
SSR = np.dot(yhat - np.mean(y), yhat - np.mean(y))

shape = np.shape(X)
#shape[0] = N, shape[1] = p+1
#df: N-p-1, df1: p
df = shape[0] - shape[1]
df1 = shape[1]-1

MSE = SSE / (df)
MSR = SSR/ (df1)
                 
F = MSR/MSE
p1 =  1-stats.f.cdf(F, df1 , df)

In [364]:
p1

1.1102230246251565e-16

In [365]:
p1<alpha

True

Result shows the p-value of F is $1.1102230246251565e-16<\alpha$ so that the Null hypothesis($H_0$) should be rejected.

(b) $H_0:\beta_3=0$ vs $H_1:\beta_3\neq0$

$ T = \frac{\beta_3-0}{\hat{\sigma}\sqrt{(X^TX)_{ii}}} \sim t(N-p-1)$

In [366]:
sig = np.sqrt(MSE)
T = (beta[3] - 0)/(sig*np.sqrt(XTXI[3,3]))
p2 = 2*(1-stats.t.cdf(abs(T), df = df))

In [367]:
p2

0.81758878828207

In [368]:
p2 < alpha

False

Result shows the p-value of $T_3$ is $0.81758878828207>\alpha$ so that the Null hypothesis($H_0$) should be accepted.